File: solutions/lua/grep/2/grep.lua
return function(options)
	local matches = setmetatable({}, { __index = table })
	local flags = table.concat(options.flags)
	local pattern = options.pattern
	if flags:match("-x") then pattern = ("^%s$"):format(pattern) end
	if flags:match("-i") then pattern = string.lower(pattern) end

	local function filegrep(filename)
		local n = 1
		for line in io.lines(filename) do
			local match_line = line
			if flags:match("-i") then match_line = match_line:lower() end
			if flags:match("-v") then
				if not match_line:match(pattern) then
					matches:insert({ filename = filename, line = n, contents = line })
				end
			else
				if match_line:match(pattern) then matches:insert({ filename = filename, line = n, contents = line }) end
			end
			n = n + 1
		end
	end

	local function result(nfiles)
		local r = setmetatable({}, { __index = table })
		for _, match in ipairs(matches) do
			local line = match.contents
			if flags:match("-n") then line = ("%d:%s"):format(match.line, line) end
			if nfiles > 1 then line = ("%s:%s"):format(match.filename, line) end
			if flags:match("-l") then line = match.filename end
			if #r == 0 or r[#r] ~= line then r:insert(line) end
		end
		return r
	end

	for _, file in ipairs(options.files) do
		filegrep(file)
	end
	return result(#options.files)
end

File: solutions/ruby/grep/2/grep.rb
# frozen_string_literal: true

# class Grep
class Grep
  def self.grep(pattern, flags, files) =
    run(files,
        build_matcher(pattern, flags),
        flags.include?('-l'),
        flags.include?('-n'),
        flags.include?('-v'),
        files.length > 1)

  def self.run(files, matcher, list_files, number_lines, invert, multiple_files)
    files.flat_map do |file|
      matches = File.foreach(file, chomp: true).with_index(1).filter_map do |line, index|
        match = matcher.match?(line)
        next if match == invert
        next file if list_files

        line = "#{index}:#{line}" if number_lines
        line = "#{file}:#{line}" if multiple_files
        line
      end
      list_files && matches.first(1) || matches
    end.join("\n")
  end

  def self.build_matcher(pattern, flags)
    source = Regexp.escape(pattern)
    source = "^#{source}$" if flags.include?('-x')
    Regexp.new(source, flags.include?('-i') ? Regexp::IGNORECASE : 0)
  end

  private_class_method :build_matcher, :run
end
